src/common/utils.ts

```diff
@@ -127,4 +127,59 @@ export function formatError(e: any): string {
 	} catch (_) {
 		return e.message;
 	}
-}
\ No newline at end of file
+}
+
+export const toPromise = Symbol();
+
+export interface PromiseAdapter<T, U> {
+	(
+		value: T,
+		resolve:
+			(value?: U | PromiseLike<U>) => void,
+		reject:
+			(reason: any) => void
+	): any;
+}
+
+declare module 'vscode' {
+	interface EventEmitter<T> {
+		/**
+		 * Return a promise that resolves with the next emitted event, or with some future
+		 * event as decided by an adapter.
+		 *
+		 * If specified, the adapter is a function that will be called with
+		 * `(event, resolve, reject)`. It will be called once per event until it resolves or
+		 * rejects.
+		 *
+		 * The default adapter is the passthrough function `(value, resolve) => resolve(value)`.
+		 *
+		 * @param {PromiseAdapter<T, U>?} adapter controls resolution of the returned promise
+		 * @returns {Promise<U>} a promise that resolves or rejects as specified by the adapter
+		 */
+		[toPromise]<U>(adapter?: PromiseAdapter<T, U>): Promise<U>;
+	}
+}
+
+const passthrough = (value, resolve) => resolve(value);
+
+EventEmitter.prototype[toPromise] = async function<T, U>(adapter=passthrough as PromiseAdapter<T, U>) {
+	let subscription;
+	return new Promise<U>((resolve, reject) =>
+		subscription = this.event((value: T) => {
+			try {
+				adapter(value, resolve, reject);
+			} catch(error) {
+				reject(error);
+			}
+		})
+	).then(
+		(result: U) => {
+			subscription.dispose();
+			return result;
+		},
+		error => {
+			subscription.dispose();
```

Comment:
Can we have a wrapper of EventEmitter instead of patching it? EventEmitter is shared in the extension host so we may want to keep it untouched.

Reply:
I thought that might be controversial 😉 

Done!